### DISCUSSION TITLE
Login synchornisation on Distributed always on AG

### DISCUSSION THREAD

**Discussion by EbenRaja** [dataplat/dbatools#7845] [Ideas]
Hello there,
I have used one of the feature put forth in article "keeping availability group logins in sync automatically" and it worked out well on AG servers, however on an DAG setup it fails when i trigger my PS script 
I am perplexed why the listener is searching for the DAG name here & wished to know if something has to be altered or the parent script wouldn't work out for an DAG requirement to synch logins 
Please advise 
----
D:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\AG\SyncLoginsToReplica.ps1 : Error while syncing logins for Availability
Group 'XXXXXX: Error connecting to [XXXXXXQAG]: The network path was not found'
At line:1 char:1
+ .\SyncLoginsToReplica.ps1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,SyncLoginsToReplica.ps1

----Script used

# define the AG name
$AvailabilityGroupName ='listenername'
 
# internal variables
    $ClientName = 'AG Login Sync helper'
    $primaryInstance = $null
    $secondaryInstances = @{}
 

try {
    # connect to the AG listener, get the name of the primary and all secondaries
        $replicas = Get-DbaAgReplica -SqlInstance $AvailabilityGroupName 
        $primaryInstance = $replicas | Where Role -eq Primary | select -ExpandProperty name
        $secondaryInstances = $replicas | Where Role -ne Primary | select -ExpandProperty name
    # create a connection object to the primary
        $primaryInstanceConnection = Connect-DbaInstance $primaryInstance -ClientName $ClientName
    # loop through each secondary replica and sync the logins
        $secondaryInstances | ForEach-Object {
            $secondaryInstanceConnection = Connect-DbaInstance $_ -ClientName $ClientName
            Copy-DbaLogin -Source $primaryInstanceConnection -Destination $secondaryInstanceConnection -ExcludeSystemLogins 
        }
}
catch {
    $msg = $_.Exception.Message
    Write-Error "Error while syncing logins for Availability Group '$($AvailabilityGroupName): $msg'"
}

**Comment by andreasjordan** [comment]
Can you run the commands in the try block one by one and tell me which command failed?

**Comment by EbenRaja** [comment]
Hi Andreas,

Thanking you for your kindness to review the concern, find below

PS C:\Windows\system32>  $replicas = Get-DbaAgReplica -SqlInstance
$AvailabilityGroupName
PS C:\Windows\system32> $primaryInstance = $replicas | Where Role -eq
Primary | select -ExpandProperty name
PS C:\Windows\system32> $secondaryInstances = $replicas | Where Role -ne
Primary | select -ExpandProperty name
PS C:\Windows\system32> $primaryInstanceConnection = Connect-DbaInstance
$primaryInstance -ClientName $ClientName
Connect-DbaInstance : Cannot bind argument to parameter 'SqlInstance'
because it is null.
At line:1 char:50
+ ... maryInstanceConnection = Connect-DbaInstance $primaryInstance -Client
...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Connect-DbaInstance],
ParameterBindingValidationException
    + FullyQualifiedErrorId :
ParameterArgumentValidationErrorNullNotAllowed,Connect-DbaInstance

PS C:\Windows\system32>

In define AG name i am issuing the Listener name as that connects both
replicas on primary DC and the DR DC is an different cluster as its an DAG
setup. Using 1.0.49 DBA tools version and got this error

With latest version of DBA tools i was getting an direct error that the DAG
name is not reachable and that was just an conventional name that we gave
as DAG name during DAG setup and hence felt helpless.

I am not an expert in SQL so ay guidance and gaps from my end then let me
know

Thanking you
Eben


# define the AG name
$AvailabilityGroupName ='XXXXXXAG01'

**Comment by EbenRaja** [comment]
Hi Andreas,

Where you are able to review the issue I am facing on Distributed always on
AG.

Appreciate your help on this regard

BR
Eben

**Comment by andreasjordan** [comment]
`$primaryInstanceConnection = Connect-DbaInstance $primaryInstance -ClientName $ClientName` is missing the parameter name `-SqlInstance`, so it should be `$primaryInstanceConnection = Connect-DbaInstance -SqlInstance $primaryInstance -ClientName $ClientName`

**Comment by EbenRaja** [comment]
Hi Andreas,
I am stuck with new error as in below
after changing the line on code snippet and running the .ps on an whole i
am now getting an different error as in below , so i tried to break the
code by running the try block statement one by one and the error now is
different

*SHPPTHPLPAG01 *is my listener name and i am able to connect this listener
from within and outside servers using SSMS tool.

PS D:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\AG> .\SyncLoginsToReplica.ps1
D:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\AG\SyncLoginsToReplica.ps1 :
Error while syncing logins for Availability
Group '*SHPPTHPLPAG01: The network path was not found*'
At line:1 char:1
+ .\SyncLoginsToReplica.ps1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error],
WriteErrorException
    + FullyQualifiedErrorId :
Microsoft.PowerShell.Commands.WriteErrorException,SyncLoginsToReplica.ps1

*Error within try block when executed one by one :*

PS D:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\AG> $replicas =
Get-DbaAgReplica -SqlInstance $AvailabilityGroupName
PS D:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\AG> $primaryInstance =
$replicas | Where Role -eq Primary | select -Expan
dProperty name
PS D:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\AG> $secondaryInstances =
$replicas | Where Role -ne Primary | select -Ex
pandProperty name
PS D:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\AG>
$primaryInstanceConnection = Connect-DbaInstance -SqlInstance $primar
yInstance -ClientName $ClientName
Connect-DbaInstance :
*Cannot bind argument to parameter 'SqlInstance' because it is null.At
line:1 char:63*
+ ... onnection = Connect-DbaInstance -SqlInstance $primaryInstance -Client
...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Connect-DbaInstance],
ParameterBindingValidationException
    + FullyQualifiedErrorId :
ParameterArgumentValidationErrorNullNotAllowed,Connect-DbaInstance

**Comment by EbenRaja** [comment]
Define and internal variable statements as in below

# define the AG name
$AvailabilityGroupName ='SHPPTHPLPAG01'

# internal variables
    $ClientName = 'AG Login Sync helper'
    $primaryInstance = $null
    $secondaryInstances = @{}

**Comment by andreasjordan** [comment]
Sorry, but I don't see clearly where the error is thrown. "The network path was not found" is a typical problem when the instance could not be found - so no problem inside of dbatools. Unfortunately, that's a little outside of what I can do on the side here.

**Comment by EbenRaja** [comment]
Thank you Andreas for your time and attention to my issues, so you confirm
to me that this has nothing to do with this script to not work properly on
an Distributed always on AG setup and it wouldn't search for the DAG name
?  as i remember somewhere it was searching for the DAG name and i am not
able to reproduce that now . I am sticking with using the listener name
only &  since that connects for me internally and externally (remote) from
server SSMS  and hence the concern.
I would work out to find an alternative solution for DAG, but I take this
time to thank you since this script works well for my standard AG setup
within the same DC.
Thanking you
Eben
